File: rgctl-tests/ecommerce-go/internal/repository/order.go
package repository

import (
	"errors"

	"github.com/sshaaf/ecommerce-go/internal/models"
	"gorm.io/gorm"
)

type OrderRepository struct {
	db *gorm.DB
}

func NewOrderRepository(db *gorm.DB) *OrderRepository {
	return &OrderRepository{db: db}
}

func (r *OrderRepository) Create(order *models.Order) error {
	return r.db.Create(order).Error
}

func (r *OrderRepository) AddItem(item *models.OrderItem) error {
	return r.db.Create(item).Error
}

func (r *OrderRepository) ListForUser(userID string) ([]models.Order, error) {
	var orders []models.Order
	err := r.db.Where("user_id = ?", userID).Order("created_at desc").Find(&orders).Error
	return orders, err
}

func (r *OrderRepository) FindByID(id string) (*models.Order, error) {
	var order models.Order
	err := r.db.First(&order, "id = ?", id).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, err
	}
	return &order, nil
}

func (r *OrderRepository) ItemsForOrder(orderID string) ([]models.OrderItem, error) {
	var items []models.OrderItem
	err := r.db.Where("order_id = ?", orderID).Find(&items).Error
	return items, err
}

File: rgctl-tests/ecommerce-rust/src/repositories/order.rs
use crate::{
    error::AppResult,
    models::order::{Order, OrderItem},
};
use sqlx::SqlitePool;

pub async fn create(pool: &SqlitePool, order: &Order) -> AppResult<Order> {
    sqlx::query_as::<_, Order>(
        "INSERT INTO orders (id, user_id, status, total_cents, created_at) VALUES (?, ?, ?, ?, ?) RETURNING id, user_id, status, total_cents, created_at",
    )
    .bind(&order.id)
    .bind(&order.user_id)
    .bind(&order.status)
    .bind(order.total_cents)
    .bind(&order.created_at)
    .fetch_one(pool)
    .await
    .map_err(Into::into)
}

pub async fn add_item(pool: &SqlitePool, item: &OrderItem) -> AppResult<()> {
    sqlx::query(
        "INSERT INTO order_items (id, order_id, product_id, quantity, unit_price_cents) VALUES (?, ?, ?, ?, ?)",
    )
    .bind(&item.id)
    .bind(&item.order_id)
    .bind(&item.product_id)
    .bind(item.quantity)
    .bind(item.unit_price_cents)
    .execute(pool)
    .await?;
    Ok(())
}

pub async fn list_for_user(pool: &SqlitePool, user_id: &str) -> AppResult<Vec<Order>> {
    sqlx::query_as::<_, Order>(
        "SELECT id, user_id, status, total_cents, created_at FROM orders WHERE user_id = ? ORDER BY created_at DESC",
    )
    .bind(user_id)
    .fetch_all(pool)
    .await
    .map_err(Into::into)
}

pub async fn find_by_id(pool: &SqlitePool, id: &str) -> AppResult<Option<Order>> {
    sqlx::query_as::<_, Order>(
        "SELECT id, user_id, status, total_cents, created_at FROM orders WHERE id = ?",
    )
    .bind(id)
    .fetch_optional(pool)
    .await
    .map_err(Into::into)
}

pub async fn items_for_order(pool: &SqlitePool, order_id: &str) -> AppResult<Vec<OrderItem>> {
    sqlx::query_as::<_, OrderItem>(
        "SELECT id, order_id, product_id, quantity, unit_price_cents FROM order_items WHERE order_id = ?",
    )
    .bind(order_id)
    .fetch_all(pool)
    .await
    .map_err(Into::into)
}
